README.pod
=pod

=encoding UTF-8

=head1 NAME

Package::Stash - Routines for manipulating stashes

=head1 VERSION

version 0.40

=head1 SYNOPSIS

  my $stash = Package::Stash->new('Foo');
  $stash->add_symbol('%foo', {bar => 1});
  # $Foo::foo{bar} == 1
  $stash->has_symbol('$foo') # false
  my $namespace = $stash->namespace;
  *{ $namespace->{foo} }{HASH} # {bar => 1}

=head1 DESCRIPTION

Manipulating stashes (Perl's symbol tables) is occasionally necessary, but
incredibly messy, and easy to get wrong. This module hides all of that behind a
simple API.

NOTE: Most methods in this class require a variable specification that includes
a sigil. If this sigil is absent, it is assumed to represent the IO slot.

Due to limitations in the typeglob API available to perl code, and to typeglob
manipulation in perl being quite slow, this module provides two
implementations - one in pure perl, and one using XS. The XS implementation is
to be preferred for most usages; the pure perl one is provided for cases where
XS modules are not a possibility. The current implementation in use can be set
by setting C<$ENV{PACKAGE_STASH_IMPLEMENTATION}> or
C<$Package::Stash::IMPLEMENTATION> before loading Package::Stash (with the
environment variable taking precedence), otherwise, it will use the XS
implementation if possible, falling back to the pure perl one.

=head1 METHODS

=head2 new $package_name

Creates a new C<Package::Stash> object, for the package given as the only
argument.

=head2 name

Returns the name of the package that this object represents.

=head2 namespace

Returns the raw stash itself.

=head2 add_symbol $variable $value %opts

Adds a new package symbol, for the symbol given as C<$variable>, and optionally
gives it an initial value of C<$value>. C<$variable> should be the name of
variable including the sigil, so

  Package::Stash->new('Foo')->add_symbol('%foo')

will create C<%Foo::foo>.

Valid options (all optional) are C<filename>, C<first_line_num>, and
C<last_line_num>.

C<$opts{filename}>, C<$opts{first_line_num}>, and C<$opts{last_line_num}> can
be used to indicate where the symbol should be regarded as having been defined.
Currently these values are only used if the symbol is a subroutine ('C<&>'
sigil) and only if C<$^P & 0x10> is true, in which case the special C<%DB::sub>
hash is updated to record the values of C<filename>, C<first_line_num>, and
C<last_line_num> for the subroutine. If these are not passed, their values are
inferred (as much as possible) from C<caller> information.

=head2 remove_glob $name

Removes all package variables with the given name, regardless of sigil.

=head2 has_symbol $variable

Returns whether or not the given package variable (including sigil) exists.

=head2 get_symbol $variable

Returns the value of the given package variable (including sigil).

=head2 get_or_add_symbol $variable

Like C<get_symbol>, except that it will return an empty hashref or
arrayref if the variable doesn't exist.

=head2 remove_symbol $variable

Removes the package variable described by C<$variable> (which includes the
sigil); other variables with the same name but different sigils will be
untouched.

=head2 list_all_symbols $type_filter

Returns a list of package variable names in the package, without sigils. If a
C<type_filter> is passed, it is used to select package variables of a given
type, where valid types are the slots of a typeglob ('SCALAR', 'CODE', 'HASH',
etc). Note that if the package contained any C<BEGIN> blocks, perl will leave
an empty typeglob in the C<BEGIN> slot, so this will show up if no filter is
used (and similarly for C<INIT>, C<END>, etc).

=head2 get_all_symbols $type_filter

Returns a hashref, keyed by the variable names in the package. If
C<$type_filter> is passed, the hash will contain every variable of that type in
the package as values, otherwise, it will contain the typeglobs corresponding
to the variable names (basically, a clone of the stash).

=for stopwords profilers

This is especially useful for debuggers and profilers, which use C<%DB::sub> to
determine where the source code for a subroutine can be found.  See
L<http://perldoc.perl.org/perldebguts.html#Debugger-Internals> for more
information about C<%DB::sub>.

=head1 WORKING WITH VARIABLES

It is important to note, that when working with scalar variables, the default
behavior is to B<copy> values.

  my $stash = Package::Stash->new('Some::Namespace');
  my $variable = 1;
  # $Some::Namespace::name is a copy of $variable
  $stash->add_symbol('$name', $variable);
  $variable++
  # $Some::Namespace::name == 1 , $variable == 2

This will likely confuse people who expect it to work the same as typeglob
assignment, which simply creates new references to existing variables.

  my $variable = 1;
  {
      no strict 'refs';
      # assign $Package::Stash::name = $variable
      *{'Package::Stash::name'} = \$variable;
  }
  $variable++ # affects both names

If this behaviour is desired when working with Package::Stash, simply pass
Package::Stash a scalar ref:

  my $stash = Package::Stash->new('Some::Namespace');
  my $variable = 1;
  # $Some::Namespace::name is now $variable
  $stash->add_symbol('$name', \$variable);
  $variable++
  # $Some::Namespace::name == 2 , $variable == 2

This will be what you want as well if you're ever working with L<Readonly>
variables:

  use Readonly;
  Readonly my $value, 'hello';

  $stash->add_symbol('$name', \$value); # reference
  print $Some::Namespace::name; # hello
  # Tries to modify the read-only 'hello' and dies.
  $Some::Namespace::name .= " world";

  $stash->add_symbol('$name', $value); # copy
  print $Some::Namespace::name; # hello
  # No problem, modifying a copy, not the original
  $Some::Namespace::name .= " world";

=head1 BUGS / CAVEATS

=over 4

=item * Prior to perl 5.10, scalar slots are only considered to exist if they are defined

This is due to a shortcoming within perl itself. See
L<perlref/Making References> point 7 for more information.

=item * GLOB and FORMAT variables are not (yet) accessible through this module.

=item * Also, see the BUGS section for the specific backends (L<Package::Stash::XS> and L<Package::Stash::PP>)

=back

=head1 SEE ALSO

=over 4

=item * L<Class::MOP::Package>

This module is a factoring out of code that used to live here

=back

=head1 HISTORY

Based on code from L<Class::MOP::Package>, by Stevan Little and the Moose
Cabal.

=head1 SUPPORT

Bugs may be submitted through L<the RT bug tracker|https://rt.cpan.org/Public/Dist/Display.html?Name=Package-Stash>
(or L<bug-Package-Stash@rt.cpan.org|mailto:bug-Package-Stash@rt.cpan.org>).

=head1 AUTHORS

=over 4

=item *

Stevan Little <stevan.little@iinteractive.com>

=item *

Jesse Luehrs <doy@tozt.net>

=back

=head1 CONTRIBUTORS

=for stopwords Karen Etheridge Carlos Lima Christian Walde Dave Rolsky Justin Hunter Kent Fredric Niko Tyni Renee Tim Bunce

=over 4

=item *

Karen Etheridge <ether@cpan.org>

=item *

Carlos Lima <carlos@multi>

=item *

Christian Walde <walde.christian@googlemail.com>

=item *

Dave Rolsky <autarch@urth.org>

=item *

Justin Hunter <justin.d.hunter@gmail.com>

=item *

Kent Fredric <kentfredric@gmail.com>

=item *

Niko Tyni <ntyni@debian.org>

=item *

Renee <reb@perl-services.de>

=item *

Tim Bunce <Tim.Bunce@pobox.com>

=back

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2022 by Jesse Luehrs.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut


Changes
Revision history for Package-Stash

{{$NEXT}}

0.40      2022-02-21 05:38:38Z
     - update bundled version of ExtUtils::HasCompiler, for new new Apple
       include paths

0.39      2020-11-22 01:26:43Z
     - updated distribution metadata and packaging
     - configuration phase now uses ExtUtils::HasCompiler

0.38      2018-12-31 03:25:07Z
     - avoid new warning in 5.21.x regarding too many arguments for
       printf/sprintf (Renée Bäcker)
     - canonical repository moved to https://github.com/moose/Package-Stash

0.37  2014-09-21
      - fix spurious warning in taint mode (Niko Tyni, #12)

0.36  2013-09-03
      - support building with PUREPERL_ONLY

0.35  2013-07-09
      - remove old deprecated api

0.34  2013-01-04
      - support anonymous stashes on newer perl versions
      - prevent harmful effects from invalid settings for
        $ENV{PACKAGE_STASH_IMPLEMENTATION}
      - switch to Module::Implementation

0.33  2011-09-28
      - add conflict on MooseX::Method::Signatures 0.36 (ether)

0.32  2011-09-05
      - bring the behavior of has_symbol for nonexistant scalars into line with
        the xs version
      - invalid package names (for instance, Foo:Bar) are not allowed
      - invalid stash entry names (anything containing ::) are not allowed

0.31  2011-08-08
      - fix ->add_symbol('$foo', qr/sdlfk/) on 5.12+
      - fix ->add_symbol('$foo', \v1.2.3) on 5.10+

0.30  2011-07-21
      - fix compiler detection in Makefile.PL (ribasushi)

0.29  2011-04-06
      - no, *really* skip the package-stash-conflict script

0.28  2011-03-29
      - META.json fixes (mst)

0.27  2011-03-27
      - also skip the package-stash-conflicts script (Father Chrysostomos)

0.26  2011-03-04
      - make the namespace cache lazy and weak, in case the stash is deleted
      - but, this doesn't work on 5.8, so disable the namespace caching
        entirely there

0.25  2011-01-25
      - make the leak tests author-only, since some smokers run release tests

0.24  2011-01-17
      - oops, i did need the Test::Requires dep

0.23  2011-01-11
      - lower perl prereq to 5.8.1 (ribasushi)
      - make the leak tests release-only, since they keep randomly failing on
        platforms i don't have access to. people are encouraged to submit
        patches for these if they are affected.

0.22  2011-01-05
      - bump ::XS dep again

0.21  2011-01-05
      - bump Package::Stash::XS dep, since a bug was fixed there

0.20  2011-01-03
      - one more fix for the Conflicts plugin

0.19  2011-01-03
      - stop manually inserting conflict checking into Makefile.PL since we're
        using the dzil Conflicts plugin now

0.18  2011-01-03
      - non-trial release

0.17-TRIAL  2011-01-03
      - use Dist::Zilla::Plugin::Conflicts rather than doing it by hand
      - silence deprecation warnings for the method renaming for now

0.16-TRIAL  2010-12-31
      - use Dist::CheckConflicts

0.15-TRIAL  2010-11-16
      - split the XS conversion out to its own dist (Package-Stash-XS), and
        convert Package::Stash into a module which loads either the XS or pure
        perl implementation, depending on what's available

0.14-TRIAL  2010-11-14
      - complete rewrite in C, for speed (this includes the vivification
        changes from earlier). should be entirely backwards compatible
        otherwise (in terms of documented api anyway).

      - methods were renamed for brevity: s/_package//

0.13  2010-10-31
      - revert the vivification changes for now, to get an actual release out
        with Test::Fatal

0.12-TRIAL  2010-10-27
      - actually include the conflict stuff in the release (bah)

0.11-TRIAL  2010-10-27
      - conflict on mx-role-withoverloading too

0.10-TRIAL  2010-10-27
      - only do the weird ISA special-casing on perl versions where it's broken

0.09-TRIAL  2010-10-27
      - clean up the vivication code a lot, make it behave more sanely
      - use Test::Fatal instead of Test::Exception (Justin Hunter)

0.08  2010-09-18
      - oops, accidentally included some experimental changes in that last
        release, that break things

0.07  2010-09-18
      - non-dev release

0.06-TRIAL  2010-08-26
      - re-enable the caching of the stash, since I can't reproduce the bug
        at all

0.05  2010-06-15
      - bump Test::More requirement for done_testing

      - update packaging stuff

0.04  2010-06-13
      - get_package_symbol now doesn't autovivify stash entries. A new method
        get_or_add_package_symbol can now be used for that behavior.

      - Update %DB::sub on add_package_symbol (Tim Bunce).

0.03  2010-05-14
      - Rename from Stash::Manip to Package::Stash

0.02  2010-05-13
      - Need to dep on Test::Exception

0.01  2010-05-12
      - Initial release


lib/Package/Stash/PP.pm
package Package::Stash::PP;
use strict;
use warnings;
# ABSTRACT: Pure perl implementation of the Package::Stash API

our $VERSION = '0.41';

use B;
use Carp qw(confess);
use Scalar::Util qw(blessed reftype weaken);
use Symbol;
# before 5.12, assigning to the ISA glob would make it lose its magical ->isa
# powers
use constant BROKEN_ISA_ASSIGNMENT => ($] < 5.012);
# before 5.10, stashes don't ever seem to drop to a refcount of zero, so
# weakening them isn't helpful
use constant BROKEN_WEAK_STASH     => ($] < 5.010);
# before 5.10, the scalar slot was always treated as existing if the
# glob existed
use constant BROKEN_SCALAR_INITIALIZATION => ($] < 5.010);
# add_method on anon stashes triggers rt.perl #1804 otherwise
# fixed in perl commit v5.13.3-70-g0fe688f
use constant BROKEN_GLOB_ASSIGNMENT => ($] < 5.013004);
# pre-5.10, ->isa lookups were cached in the ::ISA::CACHE:: slot
use constant HAS_ISA_CACHE => ($] < 5.010);

=head1 SYNOPSIS

  use Package::Stash;

=head1 DESCRIPTION

This is a backend for L<Package::Stash> implemented in pure perl, for those without a compiler or who would like to use this inline in scripts.

=cut

sub new {
    my $class = shift;
    my ($package) = @_;

    if (!defined($package) || (ref($package) && reftype($package) ne 'HASH')) {
        confess "Package::Stash->new must be passed the name of the "
              . "package to access";
    }
    elsif (ref($package) && reftype($package) eq 'HASH') {
        confess "The PP implementation of Package::Stash does not support "
              . "anonymous stashes before perl 5.14"
            if BROKEN_GLOB_ASSIGNMENT;

        return bless {
            'namespace' => $package,
        }, $class;
    }
    elsif ($package =~ /\A[0-9A-Z_a-z]+(?:::[0-9A-Z_a-z]+)*\z/) {
        return bless {
            'package' => $package,
        }, $class;
    }
    else {
        confess "$package is not a module name";
    }

}

sub name {
    confess "Can't call name as a class method"
        unless blessed($_[0]);
    confess "Can't get the name of an anonymous package"
        unless defined($_[0]->{package});
    return $_[0]->{package};
}

sub namespace {
    confess "Can't call namespace as a class method"
        unless blessed($_[0]);

    if (BROKEN_WEAK_STASH) {
        no strict 'refs';
        return \%{$_[0]->name . '::'};
    }
    else {
        return $_[0]->{namespace} if defined $_[0]->{namespace};

        {
            no strict 'refs';
            $_[0]->{namespace} = \%{$_[0]->name . '::'};
        }

        weaken($_[0]->{namespace});

        return $_[0]->{namespace};
    }
}

{
    my %SIGIL_MAP = (
        '$' => 'SCALAR',
        '@' => 'ARRAY',
        '%' => 'HASH',
        '&' => 'CODE',
        ''  => 'IO',
    );

    sub _deconstruct_variable_name {
        my ($variable) = @_;

        my @ret;
        if (ref($variable) eq 'HASH') {
            @ret = @{$variable}{qw[name sigil type]};
        }
        else {
            (defined $variable && length $variable)
                || confess "You must pass a variable name";

            my $sigil = substr($variable, 0, 1, '');

            if (exists $SIGIL_MAP{$sigil}) {
                @ret = ($variable, $sigil, $SIGIL_MAP{$sigil});
            }
            else {
                @ret = ("${sigil}${variable}", '', $SIGIL_MAP{''});
            }
        }

        # XXX in pure perl, this will access things in inner packages,
        # in xs, this will segfault - probably look more into this at
        # some point
        ($ret[0] !~ /::/)
            || confess "Variable names may not contain ::";

        return @ret;
    }
}

sub _valid_for_type {
    my ($value, $type) = @_;
    if ($type eq 'HASH' || $type eq 'ARRAY'
     || $type eq 'IO'   || $type eq 'CODE') {
        return reftype($value) eq $type;
    }
    else {
        my $ref = reftype($value);
        return !defined($ref) || $ref eq 'SCALAR' || $ref eq 'REF' || $ref eq 'LVALUE' || $ref eq 'REGEXP' || $ref eq 'VSTRING';
    }
}

sub add_symbol {
    my ($self, $variable, $initial_value, %opts) = @_;

    my ($name, $sigil, $type) = _deconstruct_variable_name($variable);

    if (@_ > 2) {
        _valid_for_type($initial_value, $type)
            || confess "$initial_value is not of type $type";

        # cheap fail-fast check for PERLDBf_SUBLINE and '&'
        if ($^P and $^P & 0x10 && $sigil eq '&') {
            my $filename = $opts{filename};
            my $first_line_num = $opts{first_line_num};

            (undef, $filename, $first_line_num) = caller
                if not defined $filename;

            my $last_line_num = $opts{last_line_num} || ($first_line_num ||= 0);

            # http://perldoc.perl.org/perldebguts.html#Debugger-Internals
            $DB::sub{$self->name . '::' . $name} = "$filename:$first_line_num-$last_line_num";
        }
    }

    if (BROKEN_GLOB_ASSIGNMENT) {
        if (@_ > 2) {
            no strict 'refs';
            no warnings 'redefine';
            *{ $self->name . '::' . $name } = ref $initial_value
                ? $initial_value : \$initial_value;
        }
        else {
            no strict 'refs';
            if (BROKEN_ISA_ASSIGNMENT && $name eq 'ISA') {
                *{ $self->name . '::' . $name };
            }
            else {
                my $undef = _undef_ref_for_type($type);
                *{ $self->name . '::' . $name } = $undef;
            }
        }
    }
    else {
        my $namespace = $self->namespace;
        {
            # using glob aliasing instead of Symbol::gensym, because otherwise,
            # magic doesn't get applied properly.
            # see <20120710063744.19360.qmail@lists-nntp.develooper.com> on p5p
            local *__ANON__:: = $namespace;
            no strict 'refs';
            no warnings 'void';
            no warnings 'once';
            *{"__ANON__::$name"};
        }

        if (@_ > 2) {
            no warnings 'redefine';
            *{ $namespace->{$name} } = ref $initial_value
                ? $initial_value : \$initial_value;
        }
        else {
            return if BROKEN_ISA_ASSIGNMENT && $name eq 'ISA';
            *{ $namespace->{$name} } = _undef_ref_for_type($type);
        }
    }
}

sub _undef_ref_for_type {
    my ($type) = @_;

    if ($type eq 'ARRAY') {
        return [];
    }
    elsif ($type eq 'HASH') {
        return {};
    }
    elsif ($type eq 'SCALAR') {
        return \undef;
    }
    elsif ($type eq 'IO') {
        return Symbol::geniosym;
    }
    elsif ($type eq 'CODE') {
        confess "Don't know how to vivify CODE variables";
    }
    else {
        confess "Unknown type $type in vivication";
    }
}

sub remove_glob {
    my ($self, $name) = @_;
    delete $self->namespace->{$name};
}

sub has_symbol {
    my ($self, $variable) = @_;

    my ($name, $sigil, $type) = _deconstruct_variable_name($variable);

    my $namespace = $self->namespace;

    return unless exists $namespace->{$name};

    my $entry_ref = \$namespace->{$name};
    if (reftype($entry_ref) eq 'GLOB') {
        if ($type eq 'SCALAR') {
            if (BROKEN_SCALAR_INITIALIZATION) {
                return defined ${ *{$entry_ref}{$type} };
            }
            else {
                my $sv = B::svref_2object($entry_ref)->SV;
                return $sv->isa('B::SV')
                    || ($sv->isa('B::SPECIAL')
                     && $B::specialsv_name[$$sv] ne 'Nullsv');
            }
        }
        else {
            return defined *{$entry_ref}{$type};
        }
    }
    else {
        # a symbol table entry can be -1 (stub), string (stub with prototype),
        # or reference (constant)
        return $type eq 'CODE';
    }
}

sub get_symbol {
    my ($self, $variable, %opts) = @_;

    my ($name, $sigil, $type) = _deconstruct_variable_name($variable);

    my $namespace = $self->namespace;

    if (!exists $namespace->{$name}) {
        if ($opts{vivify}) {
            $self->add_symbol($variable);
        }
        else {
            return undef;
        }
    }

    my $entry_ref = \$namespace->{$name};

    if (ref($entry_ref) eq 'GLOB') {
        return *{$entry_ref}{$type};
    }
    else {
        if ($type eq 'CODE') {
            if (BROKEN_GLOB_ASSIGNMENT || defined($self->{package})) {
                no strict 'refs';
                return \&{ $self->name . '::' . $name };
            }

            # XXX we should really be able to support arbitrary anonymous
            # stashes here... (not just via Package::Anon)
            if (blessed($namespace) && $namespace->isa('Package::Anon')) {
                # ->can will call gv_init for us, which inflates the glob
                # don't know how to do this in general
                $namespace->bless(\(my $foo))->can($name);
            }
            else {
                confess "Don't know how to inflate a " . ref($entry_ref)
                      . " into a full coderef (perhaps you could use"
                      . " Package::Anon instead of a bare stash?)"
            }

            return *{ $namespace->{$name} }{CODE};
        }
        else {
            return undef;
        }
    }
}

sub get_or_add_symbol {
    my $self = shift;
    $self->get_symbol(@_, vivify => 1);
}

sub remove_symbol {
    my ($self, $variable) = @_;

    my ($name, $sigil, $type) = _deconstruct_variable_name($variable);

    # FIXME:
    # no doubt this is grossly inefficient and
    # could be done much easier and faster in XS

    my %desc = (
        SCALAR => { sigil => '$', type => 'SCALAR', name => $name },
        ARRAY  => { sigil => '@', type => 'ARRAY',  name => $name },
        HASH   => { sigil => '%', type => 'HASH',   name => $name },
        CODE   => { sigil => '&', type => 'CODE',   name => $name },
        IO     => { sigil => '',  type => 'IO',     name => $name },
    );
    confess "This should never ever ever happen" if !$desc{$type};

    my @types_to_store = grep { $type ne $_ && $self->has_symbol($desc{$_}) }
                              keys %desc;
    my %values = map { $_, $self->get_symbol($desc{$_}) } @types_to_store;

    $values{SCALAR} = $self->get_symbol($desc{SCALAR})
      if !defined $values{SCALAR}
        && $type ne 'SCALAR'
        && BROKEN_SCALAR_INITIALIZATION;

    $self->remove_glob($name);

    $self->add_symbol($desc{$_} => $values{$_})
        for grep { defined $values{$_} } keys %values;
}

sub list_all_symbols {
    my ($self, $type_filter) = @_;

    my $namespace = $self->namespace;
    if (HAS_ISA_CACHE) {
        return grep { $_ ne '::ISA::CACHE::' } keys %{$namespace}
            unless defined $type_filter;
    }
    else {
        return keys %{$namespace}
            unless defined $type_filter;
    }

    # NOTE:
    # or we can filter based on
    # type (SCALAR|ARRAY|HASH|CODE)
    if ($type_filter eq 'CODE') {
        return grep {
            # any non-typeglob in the symbol table is a constant or stub
            ref(\$namespace->{$_}) ne 'GLOB'
                # regular subs are stored in the CODE slot of the typeglob
                || defined(*{$namespace->{$_}}{CODE})
        } keys %{$namespace};
    }
    elsif ($type_filter eq 'SCALAR') {
        return grep {
            !(HAS_ISA_CACHE && $_ eq '::ISA::CACHE::') &&
            (BROKEN_SCALAR_INITIALIZATION
                ? (ref(\$namespace->{$_}) eq 'GLOB'
                      && defined(${*{$namespace->{$_}}{'SCALAR'}}))
                : (do {
                      my $entry = \$namespace->{$_};
                      ref($entry) eq 'GLOB'
                          && B::svref_2object($entry)->SV->isa('B::SV')
                  }))
        } keys %{$namespace};
    }
    else {
        return grep {
            ref(\$namespace->{$_}) eq 'GLOB'
                && defined(*{$namespace->{$_}}{$type_filter})
        } keys %{$namespace};
    }
}

sub get_all_symbols {
    my ($self, $type_filter) = @_;

    my $namespace = $self->namespace;
    return { %{$namespace} } unless defined $type_filter;

    return {
        map { $_ => $self->get_symbol({name => $_, type => $type_filter}) }
            $self->list_all_symbols($type_filter)
    }
}

=head1 BUGS

=for stopwords TODO

=over 4

=item * remove_symbol also replaces the associated typeglob

This can cause unexpected behavior when doing manipulation at compile time -
removing subroutines will still allow them to be called from within the package
as subroutines (although they will not be available as methods). This can be
considered a feature in some cases (this is how L<namespace::clean> works, for
instance), but should not be relied upon - use C<remove_glob> directly if you
want this behavior.

=item * Some minor memory leaks

The pure perl implementation has a couple minor memory leaks (see the TODO
tests in t/20-leaks.t) that I'm having a hard time tracking down - these may be
core perl bugs, it's hard to tell.

=back

=head1 SEE ALSO

=over 4

=item * L<Class::MOP::Package>

This module is a factoring out of code that used to live here

=back

=head1 AUTHOR

Mostly copied from code from L<Class::MOP::Package>, by Stevan Little and the
Moose Cabal.

=begin Pod::Coverage

BROKEN_ISA_ASSIGNMENT
add_symbol
get_all_symbols
get_or_add_symbol
get_symbol
has_symbol
list_all_symbols
name
namespace
new
remove_glob

=end Pod::Coverage

=cut


lib/Package/Stash.pm
package Package::Stash;
use strict;
use warnings;
use 5.008001;
# ABSTRACT: Routines for manipulating stashes

our $VERSION = '0.41';
our $IMPLEMENTATION;

use Module::Implementation 0.06;

BEGIN {
    local $ENV{PACKAGE_STASH_IMPLEMENTATION} = $IMPLEMENTATION
      if ( $IMPLEMENTATION and not $ENV{PACKAGE_STASH_IMPLEMENTATION} );

    Module::Implementation::build_loader_sub(
        implementations => [ 'XS', 'PP' ],
        symbols         => [qw(
            new
            name
            namespace
            add_symbol
            remove_glob
            has_symbol
            get_symbol
            get_or_add_symbol
            remove_symbol
            list_all_symbols
            get_all_symbols
        )],
    )->();
    $IMPLEMENTATION = Module::Implementation::implementation_for(__PACKAGE__);
}

=head1 SYNOPSIS

  my $stash = Package::Stash->new('Foo');
  $stash->add_symbol('%foo', {bar => 1});
  # $Foo::foo{bar} == 1
  $stash->has_symbol('$foo') # false
  my $namespace = $stash->namespace;
  *{ $namespace->{foo} }{HASH} # {bar => 1}

=head1 DESCRIPTION

Manipulating stashes (Perl's symbol tables) is occasionally necessary, but
incredibly messy, and easy to get wrong. This module hides all of that behind a
simple API.

NOTE: Most methods in this class require a variable specification that includes
a sigil. If this sigil is absent, it is assumed to represent the IO slot.

Due to limitations in the typeglob API available to perl code, and to typeglob
manipulation in perl being quite slow, this module provides two
implementations - one in pure perl, and one using XS. The XS implementation is
to be preferred for most usages; the pure perl one is provided for cases where
XS modules are not a possibility. The current implementation in use can be set
by setting C<$ENV{PACKAGE_STASH_IMPLEMENTATION}> or
C<$Package::Stash::IMPLEMENTATION> before loading Package::Stash (with the
environment variable taking precedence), otherwise, it will use the XS
implementation if possible, falling back to the pure perl one.

=method new $package_name

Creates a new C<Package::Stash> object, for the package given as the only
argument.

=method name

Returns the name of the package that this object represents.

=method namespace

Returns the raw stash itself.

=method add_symbol $variable $value %opts

Adds a new package symbol, for the symbol given as C<$variable>, and optionally
gives it an initial value of C<$value>. C<$variable> should be the name of
variable including the sigil, so

  Package::Stash->new('Foo')->add_symbol('%foo')

will create C<%Foo::foo>.

Valid options (all optional) are C<filename>, C<first_line_num>, and
C<last_line_num>.

C<$opts{filename}>, C<$opts{first_line_num}>, and C<$opts{last_line_num}> can
be used to indicate where the symbol should be regarded as having been defined.
Currently these values are only used if the symbol is a subroutine ('C<&>'
sigil) and only if C<$^P & 0x10> is true, in which case the special C<%DB::sub>
hash is updated to record the values of C<filename>, C<first_line_num>, and
C<last_line_num> for the subroutine. If these are not passed, their values are
inferred (as much as possible) from C<caller> information.

=for stopwords profilers

This is especially useful for debuggers and profilers, which use C<%DB::sub> to
determine where the source code for a subroutine can be found.  See
L<http://perldoc.perl.org/perldebguts.html#Debugger-Internals> for more
information about C<%DB::sub>.

=method remove_glob $name

Removes all package variables with the given name, regardless of sigil.

=method has_symbol $variable

Returns whether or not the given package variable (including sigil) exists.

=method get_symbol $variable

Returns the value of the given package variable (including sigil).

=method get_or_add_symbol $variable

Like C<get_symbol>, except that it will return an empty hashref or
arrayref if the variable doesn't exist.

=method remove_symbol $variable

Removes the package variable described by C<$variable> (which includes the
sigil); other variables with the same name but different sigils will be
untouched.

=method list_all_symbols $type_filter

Returns a list of package variable names in the package, without sigils. If a
C<type_filter> is passed, it is used to select package variables of a given
type, where valid types are the slots of a typeglob ('SCALAR', 'CODE', 'HASH',
etc). Note that if the package contained any C<BEGIN> blocks, perl will leave
an empty typeglob in the C<BEGIN> slot, so this will show up if no filter is
used (and similarly for C<INIT>, C<END>, etc).

=method get_all_symbols $type_filter

Returns a hashref, keyed by the variable names in the package. If
C<$type_filter> is passed, the hash will contain every variable of that type in
the package as values, otherwise, it will contain the typeglobs corresponding
to the variable names (basically, a clone of the stash).

=head1 WORKING WITH VARIABLES

It is important to note, that when working with scalar variables, the default
behavior is to B<copy> values.

  my $stash = Package::Stash->new('Some::Namespace');
  my $variable = 1;
  # $Some::Namespace::name is a copy of $variable
  $stash->add_symbol('$name', $variable);
  $variable++
  # $Some::Namespace::name == 1 , $variable == 2

This will likely confuse people who expect it to work the same as typeglob
assignment, which simply creates new references to existing variables.

  my $variable = 1;
  {
      no strict 'refs';
      # assign $Package::Stash::name = $variable
      *{'Package::Stash::name'} = \$variable;
  }
  $variable++ # affects both names

If this behaviour is desired when working with Package::Stash, simply pass
Package::Stash a scalar ref:

  my $stash = Package::Stash->new('Some::Namespace');
  my $variable = 1;
  # $Some::Namespace::name is now $variable
  $stash->add_symbol('$name', \$variable);
  $variable++
  # $Some::Namespace::name == 2 , $variable == 2

This will be what you want as well if you're ever working with L<Readonly>
variables:

  use Readonly;
  Readonly my $value, 'hello';

  $stash->add_symbol('$name', \$value); # reference
  print $Some::Namespace::name; # hello
  # Tries to modify the read-only 'hello' and dies.
  $Some::Namespace::name .= " world";

  $stash->add_symbol('$name', $value); # copy
  print $Some::Namespace::name; # hello
  # No problem, modifying a copy, not the original
  $Some::Namespace::name .= " world";

=head1 BUGS / CAVEATS

=over 4

=item * Prior to perl 5.10, scalar slots are only considered to exist if they are defined

This is due to a shortcoming within perl itself. See
L<perlref/Making References> point 7 for more information.

=item * GLOB and FORMAT variables are not (yet) accessible through this module.

=item * Also, see the BUGS section for the specific backends (L<Package::Stash::XS> and L<Package::Stash::PP>)

=back

=head1 SEE ALSO

=over 4

=item * L<Class::MOP::Package>

This module is a factoring out of code that used to live here

=back

=head1 HISTORY

Based on code from L<Class::MOP::Package>, by Stevan Little and the Moose
Cabal.

=cut
